### DISCUSSION TITLE
How to get the VK_ virtual key code constants / enum?

### DISCUSSION THREAD

**Discussion by RayKoopa** [microsoft/CsWin32#849] [Q&A] [answered]
Simple question probably

I can't figure out how to get constants / enums for `VK_RETURN`, `VK_LEFT`, `VK_A` etc. (used as `wParam` for `SendMessage(hWnd, WM_KEYDOWN, ...)`, for example).

In my `NativeMethods.txt`, I added
```
SendMessage
VK_*
WM_KEYDOWN
```
...but all it gives me are some special keys like `VK_ABNT_C2`, nothing of the usual stuff.

What am I missing?

**Comment by AArnott** [comment, marked as answer]
You're very close. The syntax you're trying with `VK_*` successfully matches constants that begin with that prefix, which is why you're getting something, though not what you want.

If you had tried `VK_RETURN` itself, your compilation would produce a warning:

> NativeMethods.txt(1,1,1,10): warning PInvoke004: Use the name of the enum that declares this constant: VIRTUAL_KEY

This warning only appears in the build output and not the error list till after your compile due to a roslyn bug. 

Anyway, that warning gives you a clue that the constant has been elevated to an enum. You should request the enum name in NativeMethods.txt and find the constant there in your C# code.

**Reply by RayKoopa** [reply]
Thanks, should've tried requesting an explicit virtual key instead of immediately going for all of them, and could've seen this :-)
